File: python/set_algorithm.py
# Implementations of set algorithms.

# Both merge and inplace_merge require seq1 and seq2 are sorted.
# And result they returned is sorted too.
def merge(seq1, seq2):
    res = []
    i = j = 0
    while i < len(seq1) and j < len(seq2) :
        if seq2[j] < seq1[i] :
            res.append(seq2[j])
            j += 1
        elif seq1[i] == seq2[j] :
            res.append(seq1[i])
            i += 1
            j += 1
        else :
            res.append(seq1[i])
            i += 1
    if i == len(seq1) : res.extend(seq2[j:])
    else : res.extend(seq1[i:])
    return res

def inplace_merge(seq1, seq2):
    j = 0
    seq2_len = len(seq2)
    for i in range(0, len(seq1)) :
        while seq1[i] >= seq2[j] and j < seq2_len :
            if seq1[i] > seq2[j] : seq1.insert(i, seq2[j])
            j += 1
    seq1.extend(seq2[j:])

def includes(seq1, seq2):
    for v in seq2 :
        if v not in seq1 : return False
    return True

def union(seq1, seq2):
    res = [v for v in seq1 if v not in seq2]
    res.extend(seq2)
    return res

def difference(seq1, seq2):
    return [v for v in seq1 if v not in seq2]

def symmetric_difference(seq1, seq2):
    return [v for v in seq1 if v not in seq2] + \
            [v for v in seq2 if v not in seq1]

def intersection(seq1, seq2):
    return [v for v in seq1 if v in seq2]

File: ruby/set_algorithm.rb
# Implementations of set algorithms.
module SetAlgorithm

# Both merge and inplace_merge require seq1 and seq2 are sorted.
# And result they returned is sorted too.
def merge(seq1, seq2)
  res = []
  i = j = 0
  while i < seq1.size && j < seq2.size
    if seq1[i] > seq2[j]
      res.push seq2[j]
      j += 1
    elsif seq1[i] == seq2[j]
      res.push seq1[i]
      i += 1
      j += 1
    else
      res.push seq1[i]
      i += 1
    end
  end
  if i == seq1.size
    res + seq2[j...seq2.size]
  else
    res + seq1[i...seq1.size]
  end
end

def inplace_merge(seq1, seq2)
  j = 0
  seq1.each do |v, i|
    while v >= seq2[j] && j < seq2.size
      seq1.insert(i, seq2[j]) if v > seq2[j]
      j += 1
    end
  end
  seq1 + seq2[j...seq2.size]
end

def includes(seq1, seq2)
  seq2.each{|v| return false unless seq1.include?(v)}
  true
end

def union(seq1, seq2)
  seq1.reject{|v| seq2.include?(v)} + seq2
end

def difference(seq1, seq2)
  seq1.reject{|v| seq2.include?(v)}
end

def symmetric_difference(seq1, seq2)
  difference(seq1, seq2) + difference(seq2, seq1)
end

def intersection(seq1, seq2)
  seq1.select{|v| seq2.include?(v)}
end

end
